presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -49,25 +49,19 @@
     private final MemoryPoolId id;
     private final long maxBytes;
 
-    @GuardedBy("this")
-    private long reservedBytes;
-    @GuardedBy("this")
-    private long reservedRevocableBytes;
+    private volatile long reservedBytes;
+    private volatile long reservedRevocableBytes;
 
     @Nullable
-    @GuardedBy("this")
     private NonCancellableMemoryFuture<?> future;
 
-    @GuardedBy("this")
     // TODO: It would be better if we just tracked QueryContexts, but their lifecycle is managed by a weak reference, so we can't do that
-    private final Map<QueryId, Long> queryMemoryReservations = new HashMap<>();
+    private final Map<QueryId, Long> queryMemoryReservations = new ConcurrentHashMap<>();
```

Comment:
All three maps (`queryMemoryReservations`, `taggedMemoryAllocations`, `queryMemoryRevocableReservations`) are here to track memory reservations per query. These maps are quite orthogonal to the main function of the `MemoryPool` class. Instead of trying to optimize locking here by applying various "ninja" techniques i would strongly recommend moving the "per query" related memory tracking to the `QueryContext`. `QueryContext` is synchronized on the per query basis, thus the locking impact should be lower. The `MemoryPool` class should remain only with two `long` fields that have to be updated under a lock (`reservedBytes`, `reservedRevocableBytes`). Updating a long under a global lock should be very lightweight operation, and thus shouldn't cause significant locking contention. 

CC @nezihyigitbasi

Reply:
Given the high risk nature of this change and very limited improvement it brings I would rather postpone it, and do the right refactoring once we have more time.